components/tikv_util/src/keybuilder.rs

```diff
@@ -0,0 +1,172 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::ptr;
+
+pub struct KeyBuilder {
+    buf: Vec<u8>,
+    start: usize,
+}
+
+impl KeyBuilder {
+    pub fn new(max_size: usize, reserved_prefix_len: usize) -> Self {
+        assert!(reserved_prefix_len < max_size);
+        let mut buf = Vec::with_capacity(max_size);
+        if reserved_prefix_len > 0 {
+            unsafe { buf.set_len(reserved_prefix_len) }
+        }
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn from_vec(
+        mut vec: Vec<u8>,
+        reserved_prefix_len: usize,
+        reserved_suffix_len: usize,
+    ) -> Self {
+        let buf = if vec.capacity() >= vec.len() + reserved_prefix_len + reserved_suffix_len {
+            if reserved_prefix_len > 0 {
+                unsafe {
+                    ptr::copy(
+                        vec.as_ptr(),
+                        vec.as_mut_ptr().add(reserved_prefix_len),
+                        vec.len(),
+                    );
+                    vec.set_len(vec.len() + reserved_prefix_len);
+                }
+            }
+            vec
+        } else {
+            let mut res = Vec::with_capacity(vec.len() + reserved_prefix_len + reserved_suffix_len);
+            if reserved_prefix_len > 0 {
+                unsafe {
+                    res.set_len(reserved_prefix_len);
+                }
+            }
+            res.extend_from_slice(vec.as_slice());
+            res
+        };
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn from_slice(s: &[u8], reserved_prefix_len: usize, reserved_suffix_len: usize) -> Self {
+        let mut buf = Vec::with_capacity(s.len() + reserved_prefix_len + reserved_suffix_len);
+        if reserved_prefix_len > 0 {
+            unsafe {
+                buf.set_len(reserved_prefix_len);
+            }
+        }
+        buf.extend_from_slice(s);
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn set_prefix(&mut self, prefix: &[u8]) {
+        assert!(self.start == prefix.len());
```

Comment:
This means we can't set it multiple times?

Reply:
To keep it simple, yes. If we need such api in the future, it is easy to change it.